### PR TITLE
Fix deploy.sh to work with oc 4.x

### PR DESCRIPTION
**Pull request by dmytro-ndp**
It's fixup of issue https://issues.jboss.org/browse/CRW-301.
Is going to test solution now.

Signed-off-by: Dmytro Nochevnov <dnochevn@redhat.com>

### CODE REVIEW

**Comment by nickboldt** [comment]
Does this still work with 3.11 or is this a breaking change to ONLY support 4.1?

**Comment by dmytro-ndp** [comment]
Is testing it with **oc 3.11** here https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/

**Comment by dmytro-ndp** [comment]
Replacement `oc create` with `oc process | oc create` doesn't work with *oc 3.11*.

**Comment by nickboldt** [comment]
I think you said "the deploy script should check which version of oc is being used, and behave differently depending on 3.x vs 4.x". But I might have heard you also say "stop supporting 3.11", which seems like an unwise solution given 3.11 isn't going away any time soon. :)

**Comment by dmytro-ndp** [comment]
:-)
I had actually meant first variant `the deploy script should check which version of oc is being used, and behave differently depending on 3.x vs 4.x`.

**Comment by nickboldt** [comment]
Good. But please remember there are much more pressing issues than making something work with oc 4 when: 

* the recommended approach for OKD/OCP/OSD 3.11 is the deploy script w/ oc 3.11
* the recommend approach for CRC/OCP/OSD 4 is operatorhub

So this really isn't something we urgently need to fix.

File: operator-installer/deploy.sh [hunk 1]
```diff
@@ -490,7 +490,7 @@ if [ ${OUT} == 0 ]; then
   printInfo "Existing operator deployment found. It will be deleted"
   ${OC_BINARY} delete deployments/codeready-operator -n=${OPENSHIFT_PROJECT} --grace-period=1 > /dev/null
 fi
-echo "${DEPLOYMENT}" | ${OC_BINARY} new-app -p IMAGE=$OPERATOR_IMAGE_NAME -n="${OPENSHIFT_PROJECT}" -f - > /dev/null
+echo "${DEPLOYMENT}" | ${OC_BINARY} process -p IMAGE=$OPERATOR_IMAGE_NAME -n="${OPENSHIFT_PROJECT}" -f - > /dev/null
```
Comment (rhopp) [inline operator-installer/deploy.sh]: shouldn't this be more like:
```
echo "${DEPLOYMENT}" | ${OC_BINARY} process -p IMAGE=$OPERATOR_IMAGE_NAME -n="${OPENSHIFT_PROJECT}" -f - | ${OC_BINARY} create -f -
```

File: operator-installer/deploy.sh [hunk 2]
```diff
@@ -512,7 +512,7 @@ createCustomResource() {
     printWarning "Custom resource codeready aleady exists. If you want the installer to create a CR, delete an existing one:"
     printWarning "${OC_BINARY} delete checlusters/codeready -n ${OPENSHIFT_PROJECT}"
   fi
-  ${OC_BINARY} new-app -f ${BASE_DIR}/custom-resource.yaml \
+  ${OC_BINARY} process -f ${BASE_DIR}/custom-resource.yaml \
```
Comment (rhopp) [inline operator-installer/deploy.sh]: The same way as previous change... replacing `new-app` with `process` does nothing... This `process` command just prints processed resource (`custom-resource.yaml` with replaced values) into standard output, which in here is sent to /dev/null :-D
You should pipe that to `oc create` like this:
```
${OC_BINARY} process -f ${BASE_DIR}/custom-resource.yaml \
               -p SERVER_IMAGE_NAME=${SERVER_IMAGE_NAME} \
               -p SERVER_IMAGE_TAG=${SERVER_IMAGE_TAG} \
               -p TLS_SUPPORT=${TLS_SUPPORT} \
               -p ENABLE_OPENSHIFT_OAUTH=${ENABLE_OPENSHIFT_OAUTH} \
               -p SELF_SIGNED_CERT=${SELF_SIGNED_CERT} \
               -n="${OPENSHIFT_PROJECT}" | oc create -f - > /dev/null
```

That's the idea and IMO it should work also on OCP/OKD 3.11 as well as on 4.1 (but needs to be tested obviously ;-))

Comment (dmytro-ndp) [reply operator-installer/deploy.sh]: Thank you for explanation. I thought `oc process -f <template> | oc create -f -` is kind of alternatives.

**Comment by dmytro-ndp** [comment]
@rhopp: your workaround works with oc 3.11 https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/88/console

**Comment by dmytro-ndp** [comment]
Fixup works for "[oc 4.1](https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest/openshift-client-linux-4.1.0.tar.gz)" as well https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/89/console

**Comment by nickboldt** [comment]
Problem 1: you're using the same job to test both OC 3.11 and 4.1. This should be either one job with two matrix configs, or two jobs. Otherwise how else can we ensure changes to deploy.sh in future are properly tested with both versions of `oc` ?

Problem 2: while the job appears to have been successful, Jenkins records it as a failure. Explain why we have a verification test that is never green.

Problem 3: the export command was removed in `oc` 4, so this needs to be fixed:

```
06:12:22 + /tmp/oc export cm/che --namespace=e2e-che
06:12:22 Error: unknown command "export" for "oc"
06:12:22 
06:12:22 Did you mean this?
06:12:22 	expose
06:12:22 
06:12:22 Run 'oc --help' for usage.
``` -- https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/89/console (using `oc` 4.1)

vs.

```
05:32:14 + /tmp/oc export cm/che --namespace=e2e-che
05:32:14 Command "export" is deprecated, use the oc get --export
05:32:14 apiVersion: v1
05:32:14 data:
05:32:14   CHE_API: http://codeready-e2e-che.apps.e2e-ocp311.crw/api
...
``` -- https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/88/console (using `oc` 3.11)

**Comment by dmytro-ndp** [comment]
@nickboldt you are right int case of dedicated job to test this changes. Besides, 
[that job]( https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP/) was used temporary to have a few builds to check installation with different oc client versions. After we will fix deploy.sh, we could consider if we need to start testing CRW with oc 4.1, to fix `oc export`, to create a matrix, etc.

**Comment by nickboldt** [comment]
If you want this merged, then YES, we need to ensure we're testing w/ oc 4.x. 

If you don't want this merged, then we can continue to just use oc 3.11.x and thus update the docs to say that deploy.sh only works with oc 3.11, because for OCP 4, there's operatorhub instead (so there's no need for this script).

**Comment by dmytro-ndp** [comment]
@nickboldt: you have requested changes in CRW CI jobs config :-)
I am not sure if this PR is a proper place for such kind of requests. Besides, I have replaced `oc export` with `oc describe` in corresponding job, and is going to fix OCP 4.x related testing jobs to use `oc 4.1` client after this PR is merged.

**Comment by nickboldt** [comment]
Merged. Verification jobs:

https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/pre-release/view/OCP%203.11/job/pre-release-upstream-che-e2e-tests-against-OCP-oc3/ >=92
https://codeready-workspaces-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/pre-release-upstream-che-e2e-tests-against-OCP-oc4/ >=3

**Pull request merged**